### ISSUE TITLE
Couldn't retrieve @Mapper annotation

### ISSUE THREAD

**Issue by remmeier** [mapstruct/mapstruct#841]
I'm currently trying to implement a processor for https://github.com/mapstruct/mapstruct/issues/702. When using the current snapshot I get:  "Couldn't retrieve @Mapper annotation". While it works fine for 1.1.0.Beta1. Maybe the given exception could be extended with more . MapperConfiguration.isValid() seems to be able to fail for various reasons. The mapper itself looks like:

@Mapper(componentModel = "cdi")
public abstract class TestMapper {

	public abstract BookEntity toEntity(Book book);
	
	@InheritInverseConfiguration(name = "toEntity")
	@Mapping(target = "ctlTcn", ignore = true)
	public abstract Book toDto(BookEntity book);

	@AfterMapping
	protected void check(Object entity, @MappingTarget Object dto) {
	}
	
	@AfterMapping
	protected void checkLockCount(Book dto, @MappingTarget BookEntity entity) {
		if (dto.getCtlTcn() != dto.getCtlTcn())
			throw new OptimisticLockException(dto.toString() + " changed...");
	}
}

**Comment by sjaakd** [comment]
Strange. Have you tried to built the entire code or only the processor? Note: all unit tests pass, and I'm pretty sure they test the '@Mapper' annotation .

**Comment by sjaakd** [comment]
@remmeier : do you have any more info on this? Did you try my suggestion above?

**Comment by remmeier** [comment]
thx, found the issue, the build was not properly working with the embedded eclipse maven distribution.

**Comment by remmeier** [comment]
BTW: the gradle daemon does seem to make life miserable when developing annotation processors with random failures like

https://discuss.gradle.org/t/custom-plugin-jar-meta-inf-gradle-plugins-properties-file-not-found-error/2499/11

and others...

**Comment by amitraju01** [comment]
I am trying implementing MapStruct for my project. Till 1.1.0.Beta1 its working nicely but not for newer versions. Every time  getting "Couldn't retrieve @Mapper annotation". Build is perfectly fine.

**Comment by agudian** [comment]
@amitraju01, could you provide a small sample project to demonstrate the issue? In a new guthub issue, please. Thanks!

**Comment by amitraju01** [comment]
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Mappings;
import org.mapstruct.factory.Mappers;

import com.anonymous.cbs.CustFinInfoInqRs;

@Mapper
    public interface CustInfoDTOMapper{
	
	CustInfoDTOMapper INSTANCE = Mappers.getMapper(CustInfoDTOMapper.class);
	@Mappings( 
	{
	 @Mapping(source = "financialClass", target = "custFinInfo.creditClassCode",qualifiedBy = FirstElement.class),
	 @Mapping(source = "systemAssignedRiskCode", target = "custFinInfo.systemRiskCode"),
	 @Mapping(source = "bankAssignedRiskCode", target = "custFinInfo.riskCode"),
.
.
})
CustFinInfoInqRs custFinInfoInqRs(CustFinInfoInqRsHostObject custFinInfoInqRsHostObject);
}
Pom.xml
 <properties>	
.
.

<org.mapstruct.version>1.1.0.CR2</org.mapstruct.version>
	</properties>

<dependency>
        <groupId>org.mapstruct</groupId>
        <artifactId>mapstruct-jdk8</artifactId> 
        <version>${org.mapstruct.version}</version>
    </dependency>
<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		     <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>3.5.1</version>
            <configuration>
                <source>1.8</source>
                <target>1.8</target> 
                <annotationProcessorPaths>
                    <path>
                        <groupId>org.mapstruct</groupId>
                        <artifactId>mapstruct-processor</artifactId>
                        <version>${org.mapstruct.version}</version>
                    </path>
                </annotationProcessorPaths>
            </configuration>
        </plugin>

While its supporting for initial releases till mapstruct-1.0.0.Final. But till mapstruct-1.0.0, nested mappings were not supported.
On maven build getting following error
ERROR] /D:/git/signature-party-service/signature-party-service/src/main/java/com/fiserv/party/custfininfo/ps/CustInfoDTOMapper.java:[12,1] Couldn't retrieve @Mapper annotation

When I comment @mapper, build runs perfectly. That means build is OK.

**Comment by amitraju01** [comment]
![capture](https://cloud.githubusercontent.com/assets/13333478/20476607/d8da2628-aff7-11e6-8893-88ee864eb368.PNG)

**Comment by amitraju01** [comment]
![capture](https://cloud.githubusercontent.com/assets/13333478/20476700/45d63974-aff8-11e6-9a65-7564ec894e0f.PNG)

**Comment by agudian** [comment]
That looks all okay so far. Can you check if it works if you disable the uber-jar cration in the spring-boot plugin.
And could you provide a small sample of the project here on github so that we can examine more closely what happens in your case?

**Comment by narvik01** [comment]
I just encountered this issue when ugrading to java 8 and there was a hidden dependency on the old mapstruct 1.0.0 artifact so both the old and mapstruct-jdk8 1.1.0 were present. Excluding the old one helped.

**Comment by filiphr** [comment]
@narvik01 thanks for the information. Can you tell us if this was from some of our own artifacts or was it from some other dependency?

@amitraju01 Can you please check if the problem is due to some dependency on the old artifact like in he case of @narvik01.

**Comment by narvik01** [comment]
@filiphr It is a dependency of http://search.maven.org/#artifactdetails|io.springfox|springfox-swagger2|2.6.1|jar

**Comment by sabarnath** [comment]
I'm still facing same issue in "*1.1.0.Final*".. Due to "springfox-swagger2|2.6.1|jar", When we expect this fix from mapstruct team...

**Comment by filiphr** [comment]
@sabarnath are you also using `mapstruct-jdk8`?

This actually is not a problem on our side. We have 2 packages, [mapstruct](http://mvnrepository.com/artifact/org.mapstruct/mapstruct) and [mapstruct-jdk8](http://mvnrepository.com/artifact/org.mapstruct/mapstruct-jdk8). The problem seems to occur when you have both on the classpath.


You don't need to exclude the entire swagger for this. You can do this to only exclude the Mapstruct dependency that swagger2 brings with itself:

```xml
<dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-swagger2</artifactId>
    <version>${swagger2.version}</version>
    <exclusions>
        <exclusion>
            <artifactId>org.mapstruct</artifactId>
            <groupId>mapstruct</groupId>
        </exclusion>
    </exclusions>
</dependency>
```

**Comment by sabarnath** [comment]
@filiphr ..We have using only on mapstruct-jdk8. Now it's working fine as per your suggestion. Thanks lot. @filiphr  and **Mapstruct team** :)

**Comment by sjaakd** [comment]
@filiphr , @sabarnath can we close this issue? 

@filiphr, @gunnarmorling : something for our FAQ (the combination swagger with mapstruct)?

**Comment by filiphr** [comment]
@sjaak it is closed already :smile:. 

I wrote on my TODO to create a ticket tonight. for our website to add this to our F&Q

**Comment by ducquoc** [comment]
Hi, I just want to add that the groupId and artifactId in Feb 15th comment should be switched: 

<dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-swagger2</artifactId>
    <version>${swagger2.version}</version>
    <exclusions>
        <exclusion>
            <groupId>org.mapstruct</groupId>
            <artifactId>mapstruct</artifactId>
        </exclusion>
    </exclusions>
</dependency>

(just a minor but I wasted like 30 mins pulling my hair out why classpath still had the old jar).

Besides, the maven-compiler-plugin version should be 3.5.1+ , and if you are using both Mapstruct and Lombok, don't put Lombok before Mapstruct.

**Comment by filiphr** [comment]
Sorry for my mistake in the comment.  There is also a question in our [FAQ](http://mapstruct.org/faq/#why-do-i-get-this-error-could-not-retrieve-mapper-annotation-during-compilation) about this and there it is correct.

Regarding Maven, in the [setup](http://mapstruct.org/documentation/installation/) for MapStruct the 3.5.1 maven compiler version is used. That should be pretty clear there.

As for Lombok and MapStruct have a look at [here](http://mapstruct.org/faq/#can-i-use-mapstruct-together-with-project-lombok). If you are using MapStruct pre 1.2.0.Beta1 and Lombok pre 1.6.14 then there is no guarantee that it will work. However, starting with MapStruct 1.2.0.Beta1  and Lombok 1.6.14 then it should work out of the box, the order of the processors is not important.

**Comment by cristian-spiescu** [comment]
Hi,

We started getting the same error on our project. The changes on our code base that induce this behavior seem pretty harmless. The issue reproduces during Maven build. The build within Eclipse works OK. I suspect it's somehow related to the fact that we have some processors that add the `@Mapper` annotation. We had an issue but it was fixed in #1045.

My question: have you got some pointers regarding where I could add a break point in the code to start digging? And/or maybe you got other hints?

Thanks in advance,
Cristian.

**Comment by filiphr** [comment]
@cristian-spiescu The entry point for MapStruct is [`MappingProcessor#L153`](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/MappingProcessor.java#L153). You can put a break point there and check all the types that are passed to the processor.

Can you check if you have both `mapstruct` and `mapstruct-jdk8` in your dependencies?
